File: AEDS 2/EXERCICIOS DE LISTA, PILHA E FILA/Pilhas/PilhaFlexivel.c
/*Exercício de pilha flexível na linguagem C. 
A pilha flexível é uma estrutura de dados que permite armazenar elementos de forma dinâmica, sem a necessidade de tamanho fixo. 
Diferente da lista, que possui várias posições tanto para inserir quanto para remover, a pilha possui apenas uma posição de inserir e remover
que se refere ao topo da pilha. Ela é uma estrutura LIFO (Last In, First Out), onde o último elemento inserido é o primeiro a ser removido. 
As operações básicas da pilha são: empilhar (push) - desempilhar (pop). */

#include <stdio.h> 
#include <stdlib.h> 

//Struct para Célula
typedef struct Celula {
    //Atributos que cada Célula vai possuir => Celula: [elemento | prox]
    int elemento; 
    struct Celula *prox; 
} Celula;

//Struct para Pilha. 
typedef struct {
    //Atributo único da pilha. 
    Celula *topo; 
} Pilha; 

//Construtor pra inicializar a pilha. 
void inicializa (Pilha *pilha) {
    //Estou me referindo que o topo da pilha aponta para NULL. 
    pilha->topo = NULL; 
}

//Função para empilhar um elemento. (Push)
void empilhar (Pilha *pilha, int x) {
    //Criar e alocar na memória uma nova célula. 
    Celula *novo = (Celula *)malloc(sizeof(Celula)); 

    //Atribuir o elemento a noa célula. 
    novo->elemento = x; 

    //Novo da pilha, vai apontar para o antigo topo. 
    novo->prox = pilha->topo; 

    //O topo passa a ser o novo elemento. 
    pilha->topo = novo; 
}

//Função para desempilhar um elemento (Pop)
int desempilhar (Pilha *pilha) {
    //Guardar o elemento removido. 
    int elementoRemovido = 0; 

    //Verificar se a pilha não está vazia. 
    if (pilha->topo != NULL) {
        //Guardar o elemento removido. 
        elementoRemovido = pilha->topo->elemento; 

        //Criar uma célula auxiliar para guardar o topo antigo.
        Celula *aux = pilha->topo; 

        //O topo passa a ser o próximo elemento do antigo topo. 
        pilha->topo = aux->prox; 

        //Liberar mémoria (apagar a célula antiga).
        free(aux); 
    }

    //Retornar o elemento removido.
    return elementoRemovido;
}

//Função para mostrar os elementos da pilha. (Mostrar)
void mostrar (Pilha *pilha) {
    //Loop para percorrer a pilha.
    for (Celula *i = pilha->topo; i != NULL; i = i->prox) {
        printf("%d ", i->elemento);
    }
    printf("\n");
}

//Main
int main() {
    //Criar uma pilha do tipo Pilha. 
    Pilha pilha; 

    //Inicializar a pilha. 
    inicializa(&pilha); 

    //Empilhar alguns elementos.
    empilhar(&pilha, 50);
    empilhar(&pilha, 40); 
    empilhar(&pilha, 30);
    empilhar(&pilha, 20);
    empilhar(&pilha, 10);

    //Mostrar os elementos da pilha.
    printf("Pilha antes de desempilhar:\n");
    mostrar(&pilha);

    //Desempilhar alguns elementos.
    desempilhar(&pilha);
    desempilhar(&pilha);
    desempilhar(&pilha);

    //Mostrar os elementos da pilha.
    printf("\nPilha após desempilhar:\n");
    mostrar(&pilha);

    return 0; 
}

File: AEDS 2/EXERCICIOS DE LISTA, PILHA E FILA/Pilhas/PilhaFlexivel.java
/*Mesmo exercício da pilha flexível porém agora em java */ 

//Criar uma classe para a célula. 
class Celula {
    //Atributos da célula. 
    int elemento; 
    Celula prox; 
}

public class PilhaFlexivel {
    //Atributo da pilha. 
    Celula topo; 

    //Método para inicializar o topo da pilha
    public PilhaFlexivel() {
        //Referindo que o topo aponta para null. 
        topo = null; 
    }

    //Método para empilhar um elemento. 
    public void empilhar (int x) {
        //Criar uma nova célula que vai ter o elemento. 
        Celula novo = new Celula(); 

        novo.elemento = x; //Atribuindo o elemento. 
        novo.prox = topo; //O topo da pilha vai apontar para o novo elemento
        topo = novo; //O novo elemento passsa a ser o topo. 
    }

    //Método para desempilhar um elemento 
    public int desempilhar() {
        //Variável para guardar o elemento removido. 
        int elementoRemovido = 0; 

        //Verificar se a pilha não está vazia
        if (topo != null) { //Para desempilhar uma pilha ela não pode estar vazia. 
            //Inidicar qual elemento vai ser removido
            elementoRemovido = topo.elemento; 

            //Criar uma célula auxiliar para guardar o elemento
            Celula aux = topo; 
            topo = aux.prox; 
        }

        //retornar o elemento. 
        return elementoRemovido; 
    }

    //Método para imprimir os elementos da pilha. 
    public void mostrar() {
        //Loop para percorrer a pilha. 
        for (Celula i = topo; i != null; i = i.prox) {
            //Imprimir
            System.out.print(i.elemento + " "); //Imprimir o elemento mais um espaço. 
        }
        System.out.println(); 
    }

    public static void main(String[] args) {
        //Criar uma pilha do tipo Pilha Flexível. 
        PilhaFlexivel pilha = new PilhaFlexivel(); 

        //Inserir elementos.
        pilha.empilhar(50); 
        pilha.empilhar(40); 
        pilha.empilhar(30); 
        pilha.empilhar(20); 
        pilha.empilhar(10); 

        //Mostrar elementos depois de inseridos. 
        System.out.print("Elemento antes das remocoes"); 
        System.out.println();
        pilha.mostrar(); 

        //Remover elementos. 
        pilha.desempilhar(); 
        pilha.desempilhar();
        pilha.desempilhar(); 

        //Mostrar elementos depois de removidos
        System.out.println();
        System.out.print("Elementos depois das remocoes: Pilha Final"); 
        System.out.println();
        pilha.mostrar(); 
    }
}